### PR TITLE
Compute allocatable for scale up from 0 based on kube-reserved

### PR DESCRIPTION
**Pull request by bskiba**

### CODE REVIEW

**Comment by bskiba** [comment]
Fixes #326

File: cluster-autoscaler/cloudprovider/gce/templates.go [hunk 1]
```diff
@@ -81,6 +85,26 @@ func (t *templateBuilder) buildCapacity(machineType string) (apiv1.ResourceList,
 	return capacity, nil
 }
 
+func (t *templateBuilder) buildAllocatable(capacity apiv1.ResourceList, kubeEnv string) (apiv1.ResourceList, error) {
```
Comment (mwielgus) [inline cluster-autoscaler/cloudprovider/gce/templates.go]: Can you (in a comment) explain what goes in in kubeEnv string?

File: cluster-autoscaler/cloudprovider/gce/templates.go [hunk 2]
```diff
@@ -126,8 +149,18 @@ func (t *templateBuilder) buildNodeFromTemplate(mig *Mig, template *gce.Instance
 				return nil, err
 			}
 			node.Spec.Taints = append(node.Spec.Taints, kubeEnvTaints...)
+
+			if allocatable, err := t.buildAllocatable(node.Status.Capacity, *item.Value); err == nil {
```
Comment (mwielgus) [inline cluster-autoscaler/cloudprovider/gce/templates.go]: what if error is not null? should we print something?

File: cluster-autoscaler/cloudprovider/gce/templates.go [hunk 3]
```diff
@@ -126,8 +149,18 @@ func (t *templateBuilder) buildNodeFromTemplate(mig *Mig, template *gce.Instance
...
+				nodeAllocatable = allocatable
+			}
 		}
 	}
+	if nodeAllocatable == nil {
+		glog.Warning("Could not extract kube-reserved from kubeEnv, setting allocatable to capacity.")
```
Comment (mwielgus) [inline cluster-autoscaler/cloudprovider/gce/templates.go]: Add information about the nodegroup for which we could not calculate allocatable

File: cluster-autoscaler/cloudprovider/gce/templates.go [hunk 4]
```diff
@@ -255,13 +331,21 @@ func extractFromKubeEnv(kubeEnv, resource string) (map[string]string, error) {
...
-					return nil, fmt.Errorf("error while parsing kube env value: %s", val)
-				}
-				result[valItems[0]] = valItems[1]
+			result = append(result, value)
+		}
+	}
+	return result, nil
+}
+
+func parseLabelsToMap(values []string) (map[string]string, error) {
```
Comment (mwielgus) [inline cluster-autoscaler/cloudprovider/gce/templates.go]: Rename - parseKeyValueList to map

File: cluster-autoscaler/cloudprovider/gce/templates.go [hunk 5]
```diff
@@ -228,20 +261,63 @@ func parseCustomMachineType(machineType string) (cpu, mem int64, err error) {
...
+func parseKubeReserved(kubeReserved string) (apiv1.ResourceList, error) {
+	resourcesMap, err := parseLabelsToMap([]string{kubeReserved})
+	if err != nil {
+		return nil, fmt.Errorf("failed to extract kube-reserved from kube-env: %q", err)
+	}
+	reservedResources := apiv1.ResourceList{}
+	for name, quantity := range resourcesMap {
+		switch apiv1.ResourceName(name) {
+		case apiv1.ResourceCPU, apiv1.ResourceMemory, apiv1.ResourceEphemeralStorage:
+			if q, err := resource.ParseQuantity(quantity); err == nil && q.Sign() >= 0 {
```
Comment (mwielgus) [inline cluster-autoscaler/cloudprovider/gce/templates.go]: what if error != nil?

File: cluster-autoscaler/cloudprovider/gce/templates.go (hunk 4 above)
Comment (bskiba) [reply cluster-autoscaler/cloudprovider/gce/templates.go]: Done

File: cluster-autoscaler/cloudprovider/gce/templates.go (hunk 3 above)
Comment (bskiba) [reply cluster-autoscaler/cloudprovider/gce/templates.go]: Done

File: cluster-autoscaler/cloudprovider/gce/templates.go (hunk 2 above)
Comment (bskiba) [reply cluster-autoscaler/cloudprovider/gce/templates.go]: This code assumes there can be multiple kube-env entries and accumulates it. If there is one error here it doesn't mean there is no allocatable at all. If this is the case, we log the Warning below. Downside is it doesn't log the error.

File: cluster-autoscaler/cloudprovider/gce/templates.go (hunk 5 above)
Comment (bskiba) [reply cluster-autoscaler/cloudprovider/gce/templates.go]: We just omit the resource which will be treated as reserved=0. Would you like me to log a warning?

**Pull request merged**